### PR TITLE
refactor fast delta profiling

### PR DESCRIPTION
**Pull request by felixge**
This PR attempts to create a better separation of concerns in the code while fixing a few bugs and improving performance along the way.

### CODE REVIEW

**Comment by pmbauer** [comment]
As this is a pretty significant rewrite, another round of fidelity check testing in relenv and testing on prod instances is warranted.  The main fastdelta branch has gone through all that and we have built some confidence in it.  Does this rewrite risk missing 1.44.0?

**Review by pmbauer — CHANGES_REQUESTED** [review]
Some error messages and comments need cleaned up, but overall I like the refactor.

Some general observations.

- Most of the memory optimizations could have been made on the original implementation.
- One thing I like about the original implementation was the readability.  Having the structure of the pprof stream un-abstracted arguably makes the code simpler and easier to comprehend as a whole.
- Abstracting the pprof parsing into `pproflite` might make this more re-usable.  With fastdelta as the only dependent, YAGNI applies. But I think making the pprof parsing more general as this refactor does is a good trade-off.

**Ideally, this near-complete-rewrite would have come after the review on https://github.com/DataDog/dd-trace-go/pull/1511 was completed and we'd already have 1.44.0 released.**

But given we've gone ahead and made bug fixes to this branch instead and effectively made this the working branch for the past week, my vote is to go ahead and get it merged.

I won't have time this week to run some production tests on this branch as we have already done with `feature-fast-delta-profiling`. So this is what I'd like to do:

1. get this merged into `feature-fast-delta` ASAP
2. get @nsrip-dd 's fix for https://github.com/DataDog/dd-trace-go/pull/1571 merged.  Nick and I chatted out-of-band, and my vote is to pay the cost of another hash pass so we separate aggregation and sample writing.  This rather than aggregating all samples in-memory then writing them out. He has a fix in progress based on this branch's refactor, another reason to merge this ASAP.  We both somehow completely missed the failed fidelity checks, mostly because it appears logs aren't being collected for the reliability env pods.  I'd not observed failure messages when searching logs, but it turns out the logs are just not there. 🤦 
3. next week, deploy `feature-fast-delta-profiling` onto a production shadow (again).
4. cut 1.44 with fastdelta disabled by default.

@nsrip-dd and I discussed #4 length out-of-band.
Even though we might have more confidence than a week ago that we've tested throughly:

- the upside is low - the performance improvement is not going to be noticed by most services
- the downside of something missed is high, forcing customers to downgrade or deploy an env var during the holidays.

Given the low upside and high downside, I do not endorse releasing fastdelta with it enabled by default on the first release.

I do not think deploying on our staging services is a good test for this case, at least not without significant effort from us.  Most of our highest-throughput services (metrics) are not actively tended to there as the teams' workflow is focused only test-in-prod techniques (shadows, canaries, etc).  Any verification and load testing in staging would be left entirely to us.

Getting teams to opt into the new feature is an easy sell if their service is dominated by delta profile resource utilization and an easy roll-back if something goes wrong.

File: profiler/internal/fastdelta/fd.go [hunk 1]
```diff
@@ -186,639 +147,210 @@ func (dc *DeltaComputer) delta(p []byte, out io.Writer) (err error) {
...
-		return fmt.Errorf("error in pruning pass: %w", err)
-	}
+	dc.encoder.Reset(out)
+	dc.decoder.Reset(p)
 
-	err = molecule.MessageEach(codec.NewBuffer(p), dc.writeStringTablePassFn(out))
-	if err != nil {
-		return fmt.Errorf("error in string table writing pass: %w", err)
+	if err := dc.pass1Index(); err != nil {
+		return fmt.Errorf("indexPass: %w", err)
```
[the remark this answers is not in the document]
Comment (felixge) [reply profiler/internal/fastdelta/fd.go]: 895b9cef

File: profiler/internal/fastdelta/fd.go [hunk 2]
```diff
@@ -186,639 +147,210 @@ func (dc *DeltaComputer) delta(p []byte, out io.Writer) (err error) {
...
+			case *pproflite.SampleType:
+				if err := dc.deltaMap.AddSampleType(t); err != nil {
+					return err
+				}
+			case *pproflite.Location:
+				dc.locationIndex.Insert(t.ID, t.Address)
+			case *pproflite.StringTable:
+				dc.strings.Add(t.Value)
+				// always include the zero-index empty string,
+				// otherwise exclude by default unless used by a kept sample in mergeSamplesPass
```
[the remark this answers is not in the document]
Comment (felixge) [reply profiler/internal/fastdelta/fd.go]: 8e74257985434724c830b3e1ac61762ab468c876

File: profiler/internal/fastdelta/fd.go [hunk 3]
```diff
@@ -186,639 +147,210 @@ func (dc *DeltaComputer) delta(p []byte, out io.Writer) (err error) {
...
+		pproflite.StringTableDecoder,
+	)
+}
+
+func (dc *DeltaComputer) pass2MergeSamples() error {
+	return dc.decoder.FieldEach(
+		func(f pproflite.Field) error {
+			sample, ok := f.(*pproflite.Sample)
+			if !ok {
+				return fmt.Errorf("indexPass: unexpected field: %T", f)
```
[the remark this answers is not in the document]
Comment (felixge) [reply profiler/internal/fastdelta/fd.go]: dd3a12f6ed506850e597298c906e92fa68fdfd06

File: profiler/internal/fastdelta/fd.go [hunk 4]
```diff
@@ -186,639 +147,210 @@ func (dc *DeltaComputer) delta(p []byte, out io.Writer) (err error) {
...
-		for _, fieldIdx := range fieldIndexes {
-			if field == fieldIdx {
-				dc.markStringIncluded(value.Number)
+func (dc *DeltaComputer) pass5WriteStringTable() error {
+	counter := 0
+	return dc.decoder.FieldEach(
+		func(f pproflite.Field) error {
+			str, ok := f.(*pproflite.StringTable)
+			if !ok {
+				return fmt.Errorf("stringTablePass: unexpected field: %T", f)
```
[the remark this answers is not in the document]
Comment (felixge) [reply profiler/internal/fastdelta/fd.go]: 12090920

File: profiler/internal/fastdelta/location_index.go [hunk 5]
```diff
@@ -54,19 +41,34 @@ func (l *locationIndex) Insert(id, address, mappingID uint64, functionIDs []uint
...
 }
 
-// GetMeta returns the mapping ID and function IDs associated with
-// the given location ID
-func (l *locationIndex) GetMeta(id uint64) (mappingID uint64, functionIDs []uint64, ok bool) {
-	loc, ok := l.get(id)
-	return loc.mappingID, l.functionIDs[loc.functionIDs.start:loc.functionIDs.end], ok
-}
-
 func (l *locationIndex) get(id uint64) (loc location, ok bool) {
```
[the remark this answers is not in the document]
Comment (felixge) [reply profiler/internal/fastdelta/location_index.go:72]: I think it's fine, but feel free to reopen if you feel strongly about it.

File: profiler/internal/pproflite/decoder.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	d := &Decoder{}
+	d.Reset(input)
+	return d
+}
+
+type Decoder struct {
+	decoders []decoder
+	input    []byte
+
+	sampleType        SampleType        // 1
```
[the remark this answers is not in the document]
Comment (felixge) [reply profiler/internal/pproflite/decoder.go:53]: 036c09a1010fc44be4ee889abb1831a8d69c88da

**Comment by felixge** [comment]
@pmbauer thank you so much for your thoughtful review and comments on this. I failed to proactively communicate on some of my work here, but the gist is that it was very much a result of my code review. I mostly wanted to make sure I understand everything very deeply and not make suggestions that I wasn't confident in being possible. It's difficult with performance sensitive code like this, because it usually is add odds with abstraction and separation of concerns.

Both Nick and I are aligned with you on shipping this to customers off by default and doing more battle hardening internally before changing that. But I also want to find some time to catch up in the next week or after (I'm mostly OOO next week) so do a mini retrospective on this project and share some more thoughts from my end and listen to yours.

Again, thank you so much for all your work on this. I'm looking forward to syncing up.

**Pull request merged**